### ISSUE TITLE
Looks like you created multiple NinjaAPIs

### ISSUE THREAD

**Issue by Raketuv** [vitalik/django-ninja#229]
Hi there, I was playing around a bit with the basic tutorial (it worked perfectly in the beginning) and got this error at some point:

```
  File "/django/django_aia_backend/django_aia_backend/urls.py", line 31, in <module>
    path('api/', include(api.urls)),
  File "/djangovenv/lib/python3.9/site-packages/ninja/main.py", line 314, in urls
    self._validate()
  File "/djangovenv/lib/python3.9/site-packages/ninja/main.py", line 398, in _validate
    raise ConfigError("\n".join(msg))
ninja.errors.ConfigError: Looks like you created multiple NinjaAPIs
To let ninja distinguish them you need to set either unique version or url_namespace
 - NinjaAPI(..., version='2.0.0')
 - NinjaAPI(..., urls_namespace='otherapi')
Already registered: ['api-1.0.0']
```

not sure what is the problem here. I only have one instantiation of NinjaAPI. I tried giving it a specific version name but that does not change anything. 

To be fair, I played around a bit before, among other things creating a django app called 'api' for using standard DRF. I deleted this and every reference to it, but even after starting a brand new django project, the error persists. Does the api instantiation get persisted somewhere in venv? I cannot explain otherwise why this happens.

**Comment by vitalik** [comment]
Hi @Raketuv 

can it be that you initiated NinjaAPI twice some how (maybe via re-import)

would be able isolate this issue into a separate zip/repo  ?

**Comment by Raketuv** [comment]
Hi @vitalik , thanks for the quick answer. I will pack it for you.
Only thing I changed now is reversed db settings to default
[Raketuv_issue_django.zip](https://github.com/vitalik/django-ninja/files/7195269/Raketuv_issue_django.zip)
.

**Comment by Raketuv** [comment]
I know what I did wrong now:
```path('api/', include(api.urls)),```

Include is not needed, that seems to be the issue here. 

My bad, thanks for the help anyway!

**Comment by Raketuv** [comment]
Hi @vitalik 
I keep getting the same error over and over again while developing other parts of my App. If I add urls in other apps, I get the error from https://github.com/vitalik/django-ninja/issues/229#issue-1000832462

To remove this, I comment out:
`path('api/', api.urls)`

I continue to debug and fix the other urls/views.

Afterwards, I put in `path('api/', api.urls)` back into the root urls.py.

Then it works again. It seems like the above error message is triggered by errors in other parts of the app, which hinders development, because it masks the true errors elsewhere. Maybe you should check the error, since in this current form I would always have to remove and add the api url during development.

**Comment by vitalik** [comment]
@Raketuv 

can it be that you reimport your NinjaAPI instance in few places ?
Just make sure you have just one import of `api` instance and include it to urls

(or maybe you can isolate your code into some zip file or repo so I can check)

**Comment by Raketuv** [comment]
@vitalik I just import it in one place. When I find the time I will provide the code for you

**Comment by vitalik** [comment]
@Raketuv 
Could you isolate this issue into separate repository or zip file so I can reproduce ?

**Comment by whit** [comment]
I have similar issue. When I have created Ninja's `TestClient()` with imported `api` instance. It works since I want to use django's `reverse()`. When I want to use `reverse` I think django utilize project `urls` file where `api` is also imported. Is there way how to avoid it? I tried to construct new `NinjaAPI` just for testsuite, but then I have got another error when tried to attach same router (`ninja.errors.ConfigError: Router@'/auth' has already been attached to API Main API:1.0`

**Comment by whit** [comment]
I'm sorry, I have found solution in mentioned PR #242. Now it works, as I added to my conftest:

```python
os.environ["NINJA_SKIP_REGISTRY"] = "yes"
```

**Comment by giova-fonseca** [comment]
maybe this could be helpful [https://django-ninja.rest-framework.com/guides/routers/]

**Comment by flaiming** [comment]
For me this error was caused by setting `NinjaAPI(csrf=False, auth=django_auth, ...)`. It was giving me similar error message as in first post, but after setting `os.environ["NINJA_SKIP_REGISTRY"] = "yes"` before the `NinjaAPI` initialization it finally showed the final cause

`ninja.errors.ConfigError: Cookie Authentication must be used with CSRF. Please use NinjaAPI(csrf=True)`

This should be shown right away in my opinion.

**Comment by eugeneyalansky** [comment]
> 
Worked for me as well. 
Spent some time before realizing the case.

Thanks @flaiming

**Comment by vitalik** [comment]
@flaiming yeah, the next version wilil solve this

**Comment by henrythc** [comment]
I encountered this error as well. In my project, we have installed imported a self-written package (https://docs.djangoproject.com/en/4.2/intro/reusable-apps/) but type the wrong file name:

`
from self_written_package.folder import wrong_file_name
`

The file does not use django-ninja but still it triggers: ninja.errors.ConfigError: Looks like you created multiple NinjaAPIs.

After we corrected the file name, it works in localhost but failed on server.

As https://github.com/vitalik/django-ninja/issues/229#issuecomment-1460258987 suggested, I added `os.environ["NINJA_SKIP_REGISTRY"] = "yes"` in order to view the real cause. But then it works both in localhost and on server...and I don't know what is going on......

**Comment by flaiming** [comment]
> @flaiming yeah, the next version will solve this

@vitalik Still not fixed in current version `0.22.2`.

**Comment by benjaoming** [comment]
I've also had to deal with this. It's very confusing since it shows up with a "tip" that doesn't work, so it's easy to spend a lot of time wondering what's going on :) It seems like [debug_server_url_reimport](https://github.com/vitalik/django-ninja/blob/cd66ac8c25162f6723768eae613e7f6ca4c2e2ca/ninja/main.py#L521) doesn't work as expected? I'm on 1.0rc0 and encountering this exact issue with Django 4.2.

I understand that we don't want multiple APIs with the same name. But since the magic detection of "development server mode" seems to have become a complex issue in itself, could it be possible to simply downgrade the exception to a warning if `DEBUG=True` ?

**Comment by DeepDiverGuy** [comment]
To those who want an easy solution, just put `os.environ["NINJA_SKIP_REGISTRY"] = "yes"` in the asgi.py file. Mine is working & I'm still in development.

**Comment by JoaoPedroMDP** [comment]
Yep, this `os.environ["NINJA_SKIP_REGISTRY"] = "yes"` made it show the real problem. But should we do this configuration? Is it safe?

**Comment by q-caron** [comment]
For anyone encountering the same issue while testing, here is a workaround for your tests which does not imply modifying `NINJA_SKIP_REGISTRY` environment variable.

As pointed out by @vitalik in his [comment above](https://github.com/vitalik/django-ninja/issues/229#issuecomment-933348249), importing your `NinjaAPI` instance elsewhere may cause the `Already registered` issue.
@whit was on the right track in [his comment ](https://github.com/vitalik/django-ninja/issues/229#issuecomment-1135830989) above saying : 

> I tried to construct new NinjaAPI just for testsuite

Unfortunately, whit and others ended up setting `NINJA_SKIP_REGISTRY` to `True`. This looks like a workaround I was not satisfied with.

➡️ Given these information and my stacktrace, I used `pytest` to isolate all resources from my project's `NinjaAPI` by managing a dedicated `NinjaAPI`, `Router` and `TestClient` (I needed that to test the authentication and permissions decorators). TLDR : use session-scoped fixtures which `yield` and `del` resources : 

```python
@pytest.fixture(scope="session")
def test_api():
    api = NinjaAPI(version="test", auth=...))
    yield api
    del api

@pytest.fixture(scope="session")
def test_router():
    router = Router()
    
    @router.get("/foo/", response=FooSchema)
    def foo(request):
        return {
            "foo"bar",
        }
    yield router
    del router

@pytest.fixture(scope="session")
def session_test_client(test_api, test_router):
    test_api.add_router("/foo/", test_router)
    client = TestClient(test_api)
    yield client
    del client
```

Enjoy!
